### DISCUSSION TITLE
Why does egui box this?

### DISCUSSION THREAD

**Discussion by auriium2** [emilk/egui#2196] [Q&A]
<img width="728" alt="image" src="https://user-images.githubusercontent.com/92664603/198861001-b3acd155-6bb3-4e92-afe6-ab850099b35b.png">

Rust newbie here. Shouldn't a normal function closure work here? Why does egui heap allocate the ui defining function? Is it because the code that modifies UI mutates it? If so, why doesn't egui expose the show_inside_dyn for clarity on its internals?

**Comment by notgull** [comment]
I believe that it's passed to an event handler, which needs to not be generic.

**Reply by auriium2** [reply]
Strange that it isn't pub exposed as a box, is this common in rust? I was under the impression that the language liked to be explicit when possible

**Reply by coderedart** [reply]
> Strange that it isn't pub exposed as a box, is this common in rust? I was under the impression that the language liked to be explicit when possible

Rust made it explicit. It won't allocate if you never call Box.

But the library `egui` chose to actively box a value and hide that decision as an implementation detail. It is not the only place you will find such inefficiencies. A good example is https://github.com/emilk/egui/pull/698 where author talks about removing string cloning for static strings (`&'static str`) to improve performance, but finally decides against it because the ergonomics loss is not worth it.

> emilk: I think the ergonomics of egui is one of its main selling points, so let us not compromise it by this much.

There are other occasions where this came up again like https://github.com/emilk/egui/issues/1004 etc.. where we sacrifice ABI friendliness for ergonomics (of both users and developers). 

In the end, Rust provides control and programmers decide what to do with that.
